include/boost/unordered/unordered_node_map.hpp

```diff
@@ -118,16 +136,22 @@ namespace boost {
         template <class A> static void destroy(A& al, value_type* p) noexcept
         {
           boost::allocator_destroy(al, p);
-          boost::allocator_deallocate(al,
-            boost::pointer_traits<
-              typename boost::allocator_pointer<A>::type>::pointer_to(*p),
-            1);
+        }
+
+        template <class A> static void destroy(A& al, init_type* p) noexcept
+        {
+          boost::allocator_destroy(al, p);
         }
 
         template <class A> static void destroy(A& al, element_type* p) noexcept
         {
           if (p->p) {
-            destroy(al,p->p);
+            using pointer_type=typename boost::allocator_pointer<A>::type;
+            using pointer_traits=boost::pointer_traits<pointer_type>;
+
+            destroy(al, p->p);
+            boost::allocator_deallocate(
+              al,pointer_traits::pointer_to(*(p->p)), 1);
```

Comment:
Why not just `boost::allocator_deallocate(al,p->p,1)`?

Reply:
Also, is not `al` supposed to be bound to `element_type`?

Reply:
`al` shouldn't be bound to `element_type` because in all actuality, we're only allocating and deallocating the `type_policy::value_type`.

If we wanted to harden these measures, we should probably add:
```cpp
BOOST_STATIC_ASSERT(
  std::is_same<
    typename boost::allocator_value_type<A>::type,
    type_policy::value_type
  >::value);
```

Reply:
I think it's ok as it's now, no need to go extra picky and do the static assert.